src/processor/operator/hash_join/join_hash_table.cpp

```diff
@@ -1,69 +1,113 @@
 #include "processor/operator/hash_join/join_hash_table.h"
 
+#include "function/comparison/comparison_functions.h"
 #include "function/hash/vector_hash_functions.h"
 
 using namespace kuzu::common;
 using namespace kuzu::storage;
+using namespace kuzu::function;
 
 namespace kuzu {
 namespace processor {
 
-JoinHashTable::JoinHashTable(MemoryManager& memoryManager, uint64_t numKeyColumns,
+JoinHashTable::JoinHashTable(MemoryManager& memoryManager,
+    std::vector<std::unique_ptr<common::LogicalType>> keyTypes,
     std::unique_ptr<FactorizedTableSchema> tableSchema)
-    : BaseHashTable{memoryManager}, numKeyColumns{numKeyColumns} {
+    : BaseHashTable{memoryManager}, keyTypes{std::move(keyTypes)} {
     auto numSlotsPerBlock = BufferPoolConstants::PAGE_256KB_SIZE / sizeof(uint8_t*);
-    assert(numSlotsPerBlock == nextPowerOfTwo(numSlotsPerBlock));
-    numSlotsPerBlockLog2 = std::log2(numSlotsPerBlock);
-    slotIdxInBlockMask = BitmaskUtils::all1sMaskForLeastSignificantBits(numSlotsPerBlockLog2);
+    initSlotConstant(numSlotsPerBlock);
     // Prev pointer is always the last column in the table.
-    colOffsetOfPrevPtrInTuple = tableSchema->getColOffset(tableSchema->getNumColumns() - 1);
+    prevPtrColOffset = tableSchema->getColOffset(tableSchema->getNumColumns() - 1);
     factorizedTable = std::make_unique<FactorizedTable>(&memoryManager, std::move(tableSchema));
+    this->tableSchema = factorizedTable->getTableSchema();
+    initFunctions();
 }
 
-bool JoinHashTable::discardNullFromKeys(
-    const std::vector<ValueVector*>& vectors, uint32_t numKeyVectors) {
+void JoinHashTable::initFunctions() {
+    entryHashFunctions.resize(keyTypes.size());
+    entryCompareFunctions.resize(keyTypes.size());
+    for (auto i = 0u; i < keyTypes.size(); ++i) {
+        getEntryHashFunction(keyTypes[i]->getPhysicalType(), entryHashFunctions[i]);
+        getEntryCompareFunction(keyTypes[i]->getPhysicalType(), entryCompareFunctions[i]);
+    }
+}
+
+static bool discardNullFromKeys(const std::vector<ValueVector*>& vectors) {
     bool hasNonNullKeys = true;
-    for (auto i = 0u; i < numKeyVectors; i++) {
-        if (!NodeIDVector::discardNull(*vectors[i])) {
+    for (auto& vector : vectors) {
+        if (!NodeIDVector::discardNull(*vector)) {
```

Comment:
Now we should move `discardNull` outside of `NodeIDVector`, as this is not limited to just node id vector.